File: backend/internal/api/handlers/doctor.go
package handlers

import (
	"context"
	"encoding/json"
	"net/http"
	"time"

	"github.com/Meheer17/ambieye/internal/db/models"
	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// DoctorHandler handles doctor-related requests
type DoctorHandler struct {
	userCollection  *mongo.Collection
	queryCollection *mongo.Collection
}

// NewDoctorHandler creates a new DoctorHandler
func NewDoctorHandler(db *mongo.Database) *DoctorHandler {
	return &DoctorHandler{
		userCollection:  db.Collection("users"),
		queryCollection: db.Collection("queries"),
	}
}

// // GetDashboard returns dashboard data for doctors
// func (h *DoctorHandler) GetDashboard(c *gin.Context) {
// 	doctorID, exists := c.Get("userID")
// 	if !exists {
// 		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found"})
// 		return
// 	}

// 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
// 	defer cancel()

// 	// Get counts of different query statuses
// 	pendingCount, err := h.queryCollection.CountDocuments(ctx, bson.M{
// 		"doctorId": doctorID,
// 		"status":   "pending",
// 	})
// 	if err != nil {
// 		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
// 		return
// 	}

// 	answeredCount, err := h.queryCollection.CountDocuments(ctx, bson.M{
// 		"doctorId": doctorID,
// 		"status":   "answered",
// 	})
// 	if err != nil {
// 		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
// 		return
// 	}

// 	// Get recent queries
// 	opts := options.Find().SetSort(bson.M{"createdAt": -1}).SetLimit(5)
// 	cursor, err := h.queryCollection.Find(ctx, bson.M{
// 		"doctorId": doctorID,
// 	}, opts)
// 	if err != nil {
// 		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
// 		return
// 	}
// 	defer cursor.Close(ctx)

// 	var recentQueries []models.Query
// 	if err := cursor.All(ctx, &recentQueries); err != nil {
// 		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
// 		return
// 	}

// 	// Prepare enhanced queries with patient information
// 	recentQueriesWithInfo := make([]models.QueryWithUserInfo, 0, len(recentQueries))
// 	for _, query := range recentQueries {
// 		// Get patient information
// 		var patient models.User
// 		err := h.userCollection.FindOne(ctx, bson.M{"_id": query.PatientID}).Decode(&patient)
// 		if err == nil { // Only add if patient found
// 			queryWithInfo := models.QueryWithUserInfo{
// 				ID:          query.ID,
// 				PatientID:   query.PatientID,
// 				PatientName: patient.FullName,
// 				DoctorID:    query.DoctorID,
// 				Question:    query.Question,
// 				Response:    query.Response,
// 				Status:      query.Status,
// 				CreatedAt:   query.CreatedAt,
// 				UpdatedAt:   query.UpdatedAt,
// 				AnsweredAt:  query.AnsweredAt,
// 			}
// 			recentQueriesWithInfo = append(recentQueriesWithInfo, queryWithInfo)
// 		}
// 	}

// 	c.JSON(http.StatusOK, gin.H{
// 		"stats": gin.H{
// 			"pendingQueries":  pendingCount,
// 			"answeredQueries": answeredCount,
// 		},
// 		"recentQueries": recentQueriesWithInfo,
// 	})
// }

// GetPatients returns all patients associated with a doctor
func (h *DoctorHandler) GetPatients(c *gin.Context) {
	doctorID, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found"})
		return
	}

	// Get doctor details
	verifyCtx, verifyCancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer verifyCancel()

	var doctor models.User
	err := h.userCollection.FindOne(verifyCtx, bson.M{"_id": doctorID}).Decode(&doctor)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"error": "Doctor not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Verify the user is actually a doctor
	if doctor.Role != "doctor" {
		c.JSON(http.StatusForbidden, gin.H{"error": "User is not a doctor"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// Find all users where doctor_id matches this user's ID
	cursor, err := h.userCollection.Find(ctx, bson.M{"doctor_id": doctor.Uuid})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}
	defer cursor.Close(ctx)

	// Decode all patients into a slice
	var patients []models.User
	if err := cursor.All(ctx, &patients); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to decode patients data"})
		return
	}

	// Convert to safe user objects to hide sensitive information
	safePatients := make([]interface{}, len(patients))
	for i, patient := range patients {
		safePatients[i] = patient
	}

	c.JSON(http.StatusOK, gin.H{"patients": safePatients})
}

// GetProfile returns the doctor's profile
func (h *DoctorHandler) GetProfile(c *gin.Context) {
	doctorID, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	var doctor models.User
	err := h.userCollection.FindOne(ctx, bson.M{"_id": doctorID}).Decode(&doctor)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"error": "Doctor not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Get query stats
	totalQueries, err := h.queryCollection.CountDocuments(ctx, bson.M{"doctorId": doctorID})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	answeredQueries, err := h.queryCollection.CountDocuments(ctx, bson.M{
		"doctorId": doctorID,
		"status":   "answered",
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"profile": doctor.ToSafeUser(),
		"stats": gin.H{
			"totalQueries":    totalQueries,
			"answeredQueries": answeredQueries,
			"responseRate":    calculateResponseRate(totalQueries, answeredQueries),
		},
	})
}

// Helper function to calculate response rate
func calculateResponseRate(total, answered int64) float64 {
	if total == 0 {
		return 0
	}
	return float64(answered) / float64(total) * 100
}

func (h *DoctorHandler) GetDashboard(c *gin.Context) {
	doctorID, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// Get counts of different query statuses
	pendingCount, err := h.queryCollection.CountDocuments(ctx, bson.M{
		"doctorId": doctorID,
		"status":   "pending",
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	answeredCount, err := h.queryCollection.CountDocuments(ctx, bson.M{
		"doctorId": doctorID,
		"status":   "answered",
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Get patient count for this doctor
	patientCount, err := h.userCollection.CountDocuments(ctx, bson.M{
		"role":     "patient",
		"doctorId": doctorID,
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Get doctor info
	var doctor models.User
	err = h.userCollection.FindOne(ctx, bson.M{"_id": doctorID}).Decode(&doctor)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Get recent queries
	opts := options.Find().SetSort(bson.M{"createdAt": -1}).SetLimit(5)
	cursor, err := h.queryCollection.Find(ctx, bson.M{
		"doctorId": doctorID,
	}, opts)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}
	defer cursor.Close(ctx)

	var recentQueries []models.Query
	if err := cursor.All(ctx, &recentQueries); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Prepare enhanced queries with patient information
	recentQueriesWithInfo := make([]models.QueryWithUserInfo, 0, len(recentQueries))
	for _, query := range recentQueries {
		// Get patient information
		var patient models.User
		err := h.userCollection.FindOne(ctx, bson.M{"_id": query.PatientID}).Decode(&patient)
		if err == nil { // Only add if patient found
			queryWithInfo := models.QueryWithUserInfo{
				ID:          query.ID,
				PatientID:   query.PatientID,
				PatientName: patient.FullName,
				DoctorID:    query.DoctorID,
				Question:    query.Question,
				Response:    query.Response,
				Status:      query.Status,
				Urgency:     query.Urgency,
				CreatedAt:   query.CreatedAt,
				UpdatedAt:   query.UpdatedAt,
				AnsweredAt:  query.AnsweredAt,
			}
			recentQueriesWithInfo = append(recentQueriesWithInfo, queryWithInfo)
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"profile": doctor.ToSafeUser(),
		"stats": gin.H{
			"pendingQueries":  pendingCount,
			"answeredQueries": answeredCount,
			"totalQueries":    pendingCount + answeredCount,
			"totalPatients":   patientCount,
		},
		"recentQueries": recentQueriesWithInfo,
	})
}

// GetPatientById returns a specific patient by ID
func (h *DoctorHandler) GetPatientById(c *gin.Context) {
	doctorID, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found"})
		return
	}

	// Get the patient ID from the URL parameter
	patientID := c.Param("id")
	if patientID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Patient ID is required"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// First verify that the user is a doctor
	var doctor models.User
	// Use doctorIDStr instead of doctorID for the database query
	err := h.userCollection.FindOne(ctx, bson.M{"_id": doctorID}).Decode(&doctor)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"error": "Doctor not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Verify the user is actually a doctor
	if doctor.Role != "doctor" {
		c.JSON(http.StatusForbidden, gin.H{"error": "User is not a doctor"})
		return
	}

	// Find the specific patient
	var patient models.User
	patienId, err := primitive.ObjectIDFromHex(patientID)
	err = h.userCollection.FindOne(ctx, bson.M{
		"_id":       patienId,
		"doctor_id": doctor.Uuid, // Ensure patient belongs to this doctor
	}).Decode(&patient)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotImplemented, gin.H{"error": "Patient not found or not assigned to you"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"patient": patient})
}

// UpdatePatientMedicalInfo updates a patient's medical information
func (h *DoctorHandler) UpdatePatientMedicalInfo(c *gin.Context) {
	doctorID, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found"})
		return
	}

	// Get the patient ID from the URL parameter
	patientID := c.Param("id")
	if patientID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Patient ID is required"})
		return
	}

	// Read the raw body to debug what's being received
	rawData, err := c.GetRawData()
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Failed to read request body"})
		return
	}

	// Create a temporary struct to handle the actual JSON structure
	var requestBody struct {
		MedicalInfo models.MedicalInfo `json:"medicalInfo"`
	}

	// Try to unmarshal into the wrapper struct
	if err := json.Unmarshal(rawData, &requestBody); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid JSON format: " + err.Error()})
		return
	}

	// Extract the medical info from the wrapper
	medicalInfo := requestBody.MedicalInfo

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// First verify that the user is a doctor
	var doctor models.User
	err = h.userCollection.FindOne(ctx, bson.M{"_id": doctorID}).Decode(&doctor)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"error": "Doctor not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Verify the user is actually a doctor
	if doctor.Role != "doctor" {
		c.JSON(http.StatusForbidden, gin.H{"error": "User is not a doctor"})
		return
	}

	// Convert patient ID to ObjectID
	patientObjID, err := primitive.ObjectIDFromHex(patientID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid patient ID format"})
		return
	}

	// Verify patient belongs to this doctor
	var patient models.User
	err = h.userCollection.FindOne(ctx, bson.M{
		"_id":       patientObjID,
		"doctor_id": doctor.Uuid,
	}).Decode(&patient)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusForbidden, gin.H{"error": "Patient not found or not assigned to you"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Update the patient's medical info
	updateResult, err := h.userCollection.UpdateOne(
		ctx,
		bson.M{"_id": patientObjID},
		bson.M{
			"$set": bson.M{
				"medicalInfo": medicalInfo,
				"updatedAt":   time.Now(),
			},
		},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update medical info"})
		return
	}

	if updateResult.ModifiedCount == 0 {
		c.JSON(http.StatusNotModified, gin.H{"message": "No changes made"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message":     "Medical info updated successfully",
		"medicalInfo": medicalInfo,
	})
}

// AddPatientVisitRecord adds a new visit record to a patient's history
func (h *DoctorHandler) AddPatientVisitRecord(c *gin.Context) {
	doctorID, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found"})
		return
	}

	// Get the patient ID from the URL parameter
	patientID := c.Param("id")
	if patientID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Patient ID is required"})
		return
	}

	// Read the raw body to debug what's being received
	rawData, err := c.GetRawData()
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Failed to read request body"})
		return
	}

	// Create a temporary struct to handle the actual JSON structure
	var requestBody struct {
		VisitRecord models.VisitRecord `json:"visitRecord"`
	}

	// Try to unmarshal into the wrapper struct
	if err := json.Unmarshal(rawData, &requestBody); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid JSON format: " + err.Error()})
		return
	}

	// Extract the visit record from the wrapper
	visitRecord := requestBody.VisitRecord

	// Set the date to now if not provided
	if visitRecord.Date.IsZero() {
		visitRecord.Date = time.Now()
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// First verify that the user is a doctor
	var doctor models.User
	err = h.userCollection.FindOne(ctx, bson.M{"_id": doctorID}).Decode(&doctor)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"error": "Doctor not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Verify the user is actually a doctor
	if doctor.Role != "doctor" {
		c.JSON(http.StatusForbidden, gin.H{"error": "User is not a doctor"})
		return
	}

	// Convert patient ID to ObjectID
	patientObjID, err := primitive.ObjectIDFromHex(patientID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid patient ID format"})
		return
	}

	// Verify patient belongs to this doctor
	var patient models.User
	err = h.userCollection.FindOne(ctx, bson.M{
		"_id":       patientObjID,
		"doctor_id": doctor.Uuid,
	}).Decode(&patient)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusForbidden, gin.H{"error": "Patient not found or not assigned to you"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Add the new visit record to the patient's records
	updateResult, err := h.userCollection.UpdateOne(
		ctx,
		bson.M{"_id": patientObjID},
		bson.M{
			"$push": bson.M{"visitRecords": visitRecord},
			"$set":  bson.M{"updatedAt": time.Now()},
		},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to add visit record"})
		return
	}

	if updateResult.ModifiedCount == 0 {
		c.JSON(http.StatusNotModified, gin.H{"message": "No changes made"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message":     "Visit record added successfully",
		"visitRecord": visitRecord,
	})
}

File: backend-php/src/api/handlers/doctor.php
<?php

namespace Api\Handlers;

use Db\Database;
use Db\Models\MedicalInfo;
use Db\Models\VisitRecord;

// Explicitly require model files to ensure they're loaded
require_once __DIR__ . '/../../db/models/MedicalInfo.php';
require_once __DIR__ . '/../../db/models/VisitRecord.php';

class DoctorHandler {
    private $pdo;

    public function __construct() {
        $this->pdo = Database::getInstance();
    }

    /**
     * Get doctor dashboard
     */
    public function getDashboard() {
        try {
            $doctorId = $_SERVER['userID'] ?? null;
            if (!$doctorId) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User ID not found']
                ];
            }

            // Get pending queries count
            $stmt = $this->pdo->prepare('SELECT COUNT(*) as count FROM queries WHERE doctorId = ? AND status = ?');
            $stmt->execute([$doctorId, 'pending']);
            $pendingCount = (int) ($stmt->fetch(\PDO::FETCH_ASSOC)['count'] ?? 0);

            // Get answered queries count
            $stmt = $this->pdo->prepare('SELECT COUNT(*) as count FROM queries WHERE doctorId = ? AND status = ?');
            $stmt->execute([$doctorId, 'answered']);
            $answeredCount = (int) ($stmt->fetch(\PDO::FETCH_ASSOC)['count'] ?? 0);

            // Get unique patients count
            $stmt = $this->pdo->prepare('SELECT COUNT(DISTINCT patientId) as count FROM queries WHERE doctorId = ?');
            $stmt->execute([$doctorId]);
            $patientCount = (int) ($stmt->fetch(\PDO::FETCH_ASSOC)['count'] ?? 0);

            // Get doctor info
            $stmt = $this->pdo->prepare('SELECT * FROM users WHERE id = ? AND role = ?');
            $stmt->execute([$doctorId, 'doctor']);
            $doctor = $stmt->fetch(\PDO::FETCH_ASSOC);

            if (!$doctor) {
                return [
                    'status' => 404,
                    'response' => ['error' => 'Doctor not found']
                ];
            }

            // Remove password and parse JSON fields
            unset($doctor['password']);
            $medicalInfoData = json_decode($doctor['medicalInfo'] ?? '{}', true);
            $doctor['medicalInfo'] = MedicalInfo::fromArray($medicalInfoData)->toArray();
            
            $visitRecordsData = json_decode($doctor['visitRecords'] ?? '[]', true);
            $doctor['visitRecords'] = array_map(function($record) {
                return VisitRecord::fromArray($record)->toArray();
            }, $visitRecordsData);

            // Get recent queries (last 5)
            $stmt = $this->pdo->prepare('
                SELECT q.*, u.fullName as patientName
                FROM queries q
                LEFT JOIN users u ON q.patientId = u.id
                WHERE q.doctorId = ?
                ORDER BY q.createdAt DESC
                LIMIT 5
            ');
            $stmt->execute([$doctorId]);
            $recentQueries = $stmt->fetchAll(\PDO::FETCH_ASSOC);

            return [
                'status' => 200,
                'response' => [
                    'profile' => $doctor,
                    'stats' => [
                        'pendingQueries' => $pendingCount,
                        'answeredQueries' => $answeredCount,
                        'totalQueries' => $pendingCount + $answeredCount,
                        'totalPatients' => $patientCount,
                    ],
                    'recentQueries' => $recentQueries,
                ]
            ];
        } catch (\Exception $e) {
            return [
                'status' => 500,
                'response' => ['error' => 'Database error']
            ];
        }
    }

    /**
     * Get all patients for a doctor
     */
    public function getPatients() {
        try {
            $doctorId = $_SERVER['userID'] ?? null;
            if (!$doctorId) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User ID not found']
                ];
            }

            // Verify doctor exists
            $stmt = $this->pdo->prepare('SELECT id FROM users WHERE id = ? AND role = ?');
            $stmt->execute([$doctorId, 'doctor']);
            $doctorRow = $stmt->fetch(\PDO::FETCH_ASSOC);

            if (!$doctorRow) {
                return [
                    'status' => 404,
                    'response' => ['error' => 'Doctor not found']
                ];
            }

            // Get all patients assigned to this doctor (doctor_id is INT, not UUID)
            $stmt = $this->pdo->prepare('SELECT * FROM users WHERE doctor_id = ? AND role = ?');
            $stmt->execute([$doctorId, 'patient']);
            $patients = $stmt->fetchAll(\PDO::FETCH_ASSOC);

            foreach ($patients as &$patient) {
                unset($patient['password']);
                $medicalInfoData = json_decode($patient['medicalInfo'] ?? '{}', true);
                $patient['medicalInfo'] = MedicalInfo::fromArray($medicalInfoData)->toArray();
                
                $visitRecordsData = json_decode($patient['visitRecords'] ?? '[]', true);
                $patient['visitRecords'] = array_map(function($record) {
                    return VisitRecord::fromArray($record)->toArray();
                }, $visitRecordsData);
            }

            return [
                'status' => 200,
                'response' => ['patients' => $patients]
            ];
        } catch (\Exception $e) {
            error_log('Error in getPatients: ' . $e->getMessage());
            return [
                'status' => 500,
                'response' => ['error' => 'Database error']
            ];
        }
    }

    /**
     * Get patient by ID
     */
    public function getPatientById($patientId) {
        try {
            $doctorId = $_SERVER['userID'] ?? null;
            if (!$doctorId) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User ID not found']
                ];
            }

            // Verify doctor exists and get UUID
            $stmt = $this->pdo->prepare('SELECT id FROM users WHERE id = ? AND role = ?');
            $stmt->execute([$doctorId, 'doctor']);
            $doctorRow = $stmt->fetch(\PDO::FETCH_ASSOC);

            if (!$doctorRow) {
                return [
                    'status' => 404,
                    'response' => ['error' => 'Doctor not found']
                ];
            }

            $doctorUuid = $doctorRow['id'];

            // Get patient and verify they belong to this doctor
            $stmt = $this->pdo->prepare('SELECT * FROM users WHERE id = ? AND role = ? AND doctor_id = ?');
            $stmt->execute([$patientId, 'patient', $doctorUuid]);
            $patient = $stmt->fetch(\PDO::FETCH_ASSOC);

            if (!$patient) {
                return [
                    'status' => 404,
                    'response' => ['error' => 'Patient not found or not assigned to you']
                ];
            }

            unset($patient['password']);
            $medicalInfoData = json_decode($patient['medicalInfo'] ?? '{}', true);
            $patient['medicalInfo'] = MedicalInfo::fromArray($medicalInfoData)->toArray();
            
            $visitRecordsData = json_decode($patient['visitRecords'] ?? '[]', true);
            $patient['visitRecords'] = array_map(function($record) {
                return VisitRecord::fromArray($record)->toArray();
            }, $visitRecordsData);

            // Get game history for the patient
            $stmt = $this->pdo->prepare('
                SELECT * FROM gameResults
                WHERE userId = ?
                ORDER BY date DESC
            ');
            $stmt->execute([$patientId]);
            $gameHistory = $stmt->fetchAll(\PDO::FETCH_ASSOC);

            // Parse JSON details in game results
            foreach ($gameHistory as &$game) {
                if ($game['details']) {
                    $game['details'] = json_decode($game['details'], true);
                }
            }

            return [
                'status' => 200,
                'response' => [
                    'patient' => $patient,
                    'gameHistory' => $gameHistory
                ]
            ];
        } catch (\Exception $e) {
            error_log('Error in getPatientById: ' . $e->getMessage());
            return [
                'status' => 500,
                'response' => ['error' => 'Database error']
            ];
        }
    }

    /**
     * Update patient medical info
     */
    public function updatePatientMedicalInfo($patientId, $body) {
        try {
            $doctorId = $_SERVER['userID'] ?? null;
            if (!$doctorId) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User ID not found']
                ];
            }

            // Extract medicalInfo from body (it comes wrapped in the request)
            $medicalInfoData = $body['medicalInfo'] ?? $body;
            if (empty($medicalInfoData)) {
                return [
                    'status' => 400,
                    'response' => ['error' => 'Medical info is required']
                ];
            }

            // Verify patient exists
            $stmt = $this->pdo->prepare('SELECT * FROM users WHERE id = ? AND role = ?');
            $stmt->execute([$patientId, 'patient']);
            if (!$stmt->fetch(\PDO::FETCH_ASSOC)) {
                return [
                    'status' => 404,
                    'response' => ['error' => 'Patient not found']
                ];
            }

            // Convert to MedicalInfo model
            $medicalInfo = MedicalInfo::fromArray($medicalInfoData);
            $medicalInfoJson = json_encode($medicalInfo->toArray());
            $now = date('Y-m-d H:i:s');

            $stmt = $this->pdo->prepare('
                UPDATE users 
                SET medicalInfo = ?, updatedAt = ?
                WHERE id = ? AND role = ?
            ');
            $stmt->execute([$medicalInfoJson, $now, $patientId, 'patient']);

            if ($stmt->rowCount() === 0) {
                return [
                    'status' => 404,
                    'response' => ['error' => 'Patient not found']
                ];
            }

            return [
                'status' => 200,
                'response' => ['message' => 'Medical info updated successfully']
            ];
        } catch (\Exception $e) {
            return [
                'status' => 500,
                'response' => ['error' => 'Database error']
            ];
        }
    }

    /**
     * Add patient visit record
     */
    public function addPatientVisitRecord($patientId, $body) {
        try {
            $doctorId = $_SERVER['userID'] ?? null;
            if (!$doctorId) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User ID not found']
                ];
            }

            // Extract visitRecord from body (it comes wrapped in the request)
            $visitRecordData = $body['visitRecord'] ?? $body;
            if (empty($visitRecordData)) {
                return [
                    'status' => 400,
                    'response' => ['error' => 'Visit record is required']
                ];
            }

            // Get current visitRecords
            $stmt = $this->pdo->prepare('SELECT visitRecords FROM users WHERE id = ? AND role = ?');
            $stmt->execute([$patientId, 'patient']);
            $result = $stmt->fetch(\PDO::FETCH_ASSOC);

            if (!$result) {
                return [
                    'status' => 404,
                    'response' => ['error' => 'Patient not found']
                ];
            }

            // Convert to VisitRecord model
            $visitRecord = VisitRecord::fromArray($visitRecordData);
            $visitRecordArray = $visitRecord->toArray();

            // Get existing visit records
            $visitRecords = json_decode($result['visitRecords'] ?? '[]', true);
            if (!is_array($visitRecords)) {
                $visitRecords = [];
            }
            
            // Add new visit record
            $visitRecords[] = $visitRecordArray;

            $now = date('Y-m-d H:i:s');
            $visitRecordsJson = json_encode($visitRecords);
            
            error_log('Updating visit record for patient ' . $patientId . ' with data: ' . $visitRecordsJson);

            $stmt = $this->pdo->prepare('
                UPDATE users
                SET visitRecords = ?, updatedAt = ?
                WHERE id = ? AND role = ?
            ');
            
            $updateResult = $stmt->execute([$visitRecordsJson, $now, $patientId, 'patient']);
            
            if (!$updateResult) {
                error_log('Update failed: ' . implode(' ', $stmt->errorInfo()));
                return [
                    'status' => 500,
                    'response' => ['error' => 'Failed to update visit record']
                ];
            }

            error_log('Visit record updated successfully. Rows affected: ' . $stmt->rowCount());

            return [
                'status' => 200,
                'response' => [
                    'message' => 'Visit record added successfully',
                    'visitRecord' => $visitRecordArray
                ]
            ];
        } catch (\Exception $e) {
            error_log('Exception in addPatientVisitRecord: ' . $e->getMessage());
            return [
                'status' => 500,
                'response' => ['error' => 'Database error: ' . $e->getMessage()]
            ];
        }
    }

    /**
     * Get doctor profile
     */
    public function getProfile() {
        try {
            $doctorId = $_SERVER['userID'] ?? null;
            if (!$doctorId) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User ID not found']
                ];
            }

            $stmt = $this->pdo->prepare('SELECT * FROM users WHERE id = ? AND role = ?');
            $stmt->execute([$doctorId, 'doctor']);
            $doctor = $stmt->fetch(\PDO::FETCH_ASSOC);

            if (!$doctor) {
                return [
                    'status' => 404,
                    'response' => ['error' => 'Doctor not found']
                ];
            }

            // Remove password
            unset($doctor['password']);

            // Parse JSON fields
            $medicalInfoData = json_decode($doctor['medicalInfo'] ?? '{}', true);
            $doctor['medicalInfo'] = MedicalInfo::fromArray($medicalInfoData)->toArray();
            
            $visitRecordsData = json_decode($doctor['visitRecords'] ?? '[]', true);
            $doctor['visitRecords'] = array_map(function($record) {
                return VisitRecord::fromArray($record)->toArray();
            }, $visitRecordsData);

            // Get query stats
            $stmt = $this->pdo->prepare('SELECT COUNT(*) as count FROM queries WHERE doctorId = ?');
            $stmt->execute([$doctorId]);
            $totalQueries = (int) ($stmt->fetch(\PDO::FETCH_ASSOC)['count'] ?? 0);

            $stmt = $this->pdo->prepare('SELECT COUNT(*) as count FROM queries WHERE doctorId = ? AND status = ?');
            $stmt->execute([$doctorId, 'answered']);
            $answeredQueries = (int) ($stmt->fetch(\PDO::FETCH_ASSOC)['count'] ?? 0);

            $responseRate = 0;
            if ($totalQueries > 0) {
                $responseRate = ($answeredQueries / $totalQueries) * 100;
            }

            return [
                'status' => 200,
                'response' => [
                    'profile' => $doctor,
                    'stats' => [
                        'totalQueries' => $totalQueries,
                        'answeredQueries' => $answeredQueries,
                        'responseRate' => $responseRate,
                    ]
                ]
            ];
        } catch (\Exception $e) {
            return [
                'status' => 500,
                'response' => ['error' => 'Database error']
            ];
        }
    }

    /**
     * Delete doctor account
     */
    public function deleteDoctor() {
        try {
            $doctorId = $_SERVER['userID'] ?? null;
            if (!$doctorId) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User ID not found']
                ];
            }

            // Delete all queries associated with this doctor (set doctorId to NULL for patient records)
            $stmt = $this->pdo->prepare('UPDATE queries SET doctorId = NULL WHERE doctorId = ?');
            $stmt->execute([$doctorId]);

            // Delete doctor account
            $stmt = $this->pdo->prepare('DELETE FROM users WHERE id = ? AND role = ?');
            $stmt->execute([$doctorId, 'doctor']);

            if ($stmt->rowCount() === 0) {
                return [
                    'status' => 404,
                    'response' => ['error' => 'Doctor not found']
                ];
            }

            return [
                'status' => 200,
                'response' => ['message' => 'Account deleted successfully']
            ];
        } catch (\Exception $e) {
            return [
                'status' => 500,
                'response' => ['error' => 'Database error']
            ];
        }
    }
}
